src/VisualStudio.Roslyn.SDK/SyntaxVisualizer/Roslyn.SyntaxVisualizer.Extension/SyntaxVisualizerContainer.xaml.cs

```diff
@@ -251,32 +255,44 @@ void IDisposable.Dispose()
         {
-            if (IsVisible && activeWpfTextView != null)
+            var cancellationToken = cancellationSeries.CreateNext();
+
+            if (!IsVisible || activeWpfTextView == null)
             {
-                var snapshot = activeWpfTextView.TextBuffer.CurrentSnapshot;
-                var contentType = snapshot.ContentType;
+                return;
+            }
 
-                if (contentType.IsOfType(VisualBasicContentType) ||
-                    contentType.IsOfType(CSharpContentType))
-                {
-                    // Get the Document corresponding to the currently active text snapshot.
-                    var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
-                    if (document != null)
-                    {
-                        // Get the SyntaxTree and SemanticModel corresponding to the Document.
-                        activeSyntaxTree = ThreadHelper.JoinableTaskFactory.Run(() => document.GetSyntaxTreeAsync());
-                        var activeSemanticModel = ThreadHelper.JoinableTaskFactory.Run(() => document.GetSemanticModelAsync());
+            var snapshot = activeWpfTextView.TextBuffer.CurrentSnapshot;
+            var contentType = snapshot.ContentType;
 
-                        // Display the SyntaxTree.
-                        if ((contentType.IsOfType(VisualBasicContentType) || contentType.IsOfType(CSharpContentType)) && activeSyntaxTree is not null)
-                        {
-                            syntaxVisualizer.DisplaySyntaxTree(document, activeSyntaxTree, activeSemanticModel, workspace: document.Project.Solution.Workspace);
-                        }
+            if (!contentType.IsOfType(VisualBasicContentType)
+                && !contentType.IsOfType(CSharpContentType))
+            {
+                return;
+            }
 
-                        NavigateFromSource();
-                    }
-                }
+            // Get the Document corresponding to the currently active text snapshot.
+            var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return;
             }
+
+            _ = ThreadHelper.JoinableTaskFactory.RunAsync(
+                async () =>
+                {
+                    // Get the SyntaxTree and SemanticModel corresponding to the Document.
+                    activeSyntaxTree = await document.GetSyntaxTreeAsync(cancellationToken);
+                    var activeSemanticModel = await document.GetSemanticModelAsync(cancellationToken);
+
+                    // Display the SyntaxTree.
+                    if (activeSyntaxTree is not null)
+                    {
+                        await syntaxVisualizer.DisplaySyntaxTreeAsync(document, activeSyntaxTree, activeSemanticModel, lazy: true, document.Project.Solution.Workspace, cancellationToken);
+                    }
+
+                    NavigateFromSource();
```

Comment:
Does it matter that this is now on a different thread? Sounds like it might interact with UI etc?

Reply:
This is still on the same thread, just at a different point in time